stream_alert/rule_processor/payload.py

```diff
@@ -0,0 +1,328 @@
+    pass
+
+
+class S3Payload(StreamPayload):
+    """S3Payload class"""
+    s3_object_size = 0
+
+    def service(self):
+        return 's3'
+
+    def pre_parse(self):
+        """Pre-parsing method for S3 objects that will download the s3 object,
+        open it for reading and iterate over lines (records) in the file.
+        This yields back references of this S3Payload instance to the caller
+        with a propertly set `pre_parsed_record` for this record.
+
+        Returns:
+            [generator] Yields instances of `self` back to the caller with the
+                proper `pre_parsed_record` set. Conforms to the interface of
+                returning a generator, providing the ability to support
+                multi-record like this (s3).
+        """
+        s3_file = self._get_object()
+        line_num, processed_size = 0, 0
+        for line_num, data in S3Payload._read_s3_file(s3_file):
+
+            self.refresh_record(data)
+            yield self
+
+            # Only do the extra calculations below if debug logging is enabled
+            if not LOGGER.isEnabledFor(log_level_debug):
+                continue
+
+            # Add the current data to the total processed size, +1 to account for line
+            # feed
+            processed_size += (len(data) + 1)
+
+            # Log a debug message on every 100 lines processed
+            if line_num and line_num % 100 == 0:
+                avg_record_size = ((processed_size - 1) / line_num)
+                if avg_record_size:
+                    approx_record_count = self.s3_object_size / avg_record_size
+                    LOGGER.debug(
+                        'Processed %s records out of an approximate total of %s '
+                        '(average record size: %s bytes, total size: %s bytes)',
+                        line_num,
+                        approx_record_count,
+                        avg_record_size,
+                        self.s3_object_size)
+
+    def _download_object(self, region, bucket, key):
+        """Download an object from S3.
+
+        Verifies the S3 object is less than or equal to 128MB, and
+        stores into a temp file.  Lambda can only execute for a
+        maximum of 300 seconds, and the file to download
+        greatly impacts that time.
+
+        Args:
+            region [string]: AWS region to use for boto client instance.
+            bucket (string): S3 bucket to download object from.
+            key [string]: Key of s3 object.
+
+        Returns:
+            [string] The downloaded path of the S3 object.
+        """
+        size_kb = self.s3_object_size / 1024.0
+        size_mb = size_kb / 1024.0
+        if size_mb > 128:
+            raise S3ObjectSizeError('S3 object to download is above 128MB')
+
+        LOGGER.debug('/tmp directory contents:%s ', os.listdir('/tmp'))
+        LOGGER.debug(os.popen('df -h /tmp | tail -1').read().strip())
+
+        if size_mb:
```

Comment:
we can condense this into one line:

`display_size = '{}MB'.format(size_mb) if size_mb else '{}KB'.format(size_kb)`